components/MainUI.tsx

```diff
@@ -1,13 +1,60 @@
-import React from 'react';
-import { Layout } from "./styles";
+import React, { useEffect, useState } from 'react';
+import { Layout } from './styles';
 import { Focus, Bar, Performer } from './';
+import Video from './Video';
 
 const { LayoutContainer } = Layout;
 
 const MainUI = () => {
+  const [participants, setParticipants] = useState([]);
+  const [room, setRoom] = useState(null);
+  const [localParticipant, setLocalParticipant] = useState(null);
+
+  const addParticipant = participant => {
+    console.log('Participant "%s" added', participant.identity);
+    setParticipants(previous => [...previous, participant]);
+  };
+
+  const removeParticipant = participant => {
+    console.log('Participant "%s" REMOVED', participant.identity);
+    setParticipants(previous =>
+      previous.filter(existing => existing.identity !== participant.identity)
+    );
+  };
+
+  const prepareRoom = room => {
+    setRoom(room);
+    setLocalParticipant(room.localParticipant);
+    window.addEventListener('beforeunload', () => room.disconnect());
+    [
+      ...Array.from(room.participants.values()),
+      room.localParticipant
+    ].forEach(participant => addParticipant(participant));
+    room.on('participantConnected', participant => {
+      console.log(`Participant connected: ${participant.identity}`);
+      addParticipant(participant);
+    });
+    room.on('participantDisconnected', participant => {
+      console.log(`Participant DISconnected: ${participant.identity}`);
+      removeParticipant(participant);
+    });
+  };
+
+  useEffect(() => {
+    fetch('/api/token')
+      .then(res => res.json())
+      .then(data => {
+        Video.connectToRoom({ token: data.token, roomName: 'cool-people-only' })
+          .then(room => prepareRoom(room))
+          .catch(err => console.log(err));
+      });
+    return () => room && room.disconnect();
```

Comment:
Returning this function means that this function will run when the component is unmounted. But wait, we're disconnecting above, too:

`window.addEventListener('beforeunload', () => room.disconnect());`

Yeah, true. I added the `beforeunload` handler because sometimes React doesn't have time to unload fully when reloading the page. We're covering all our bases here.